lib/Twig/Environment.php

```diff
@@ -887,23 +889,22 @@ public function getRuntime($class)
     public function addExtension(Twig_ExtensionInterface $extension)
     {
         $class = get_class($extension);
+        $legacyName = $class !== $extension->getName() ? $extension->getName() : null;
 
         if ($this->extensionInitialized) {
-            throw new LogicException(sprintf('Unable to register extension "%s" as extensions have already been initialized.', $class));
+            throw new LogicException(sprintf('Unable to register extension "%s" as extensions have already been initialized.', null !== $legacyName ? $legacyName : $class));
         }
 
-        $m = new ReflectionMethod($extension, 'getName');
-        $legacyName = 'Twig_Extension' !== $m->getDeclaringClass()->getName() ? $extension->getName() : null;
+        if (null !== $legacyName) {
+            if (isset($this->legacyExtensions[$legacyName])) {
+                unset($this->extensions[$class], $this->legacyExtensions[$legacyName]);
+                @trigger_error(sprintf('The possibility to register the same extension twice ("%s") is deprecated since version 1.23 and will be removed in Twig 2.0. Use proper PHP inheritance instead.', $legacyName), E_USER_DEPRECATED);
```

Comment:
It is now always set, we maintain two arrays in parallel. If you mean the same class once registered as a class and once as a legacy name, I don't think we need to support that.